coatnet/models/blocks.py

```diff
@@ -22,3 +27,155 @@ def conv_3x3(input_layer, num_filters: int = 64, downsample: bool = False):
+        strides: int,
+        conv_shortcut: bool,
+        expansion: int,
+        dropout_rate: float,
+        **kwargs
+    ):
+        super(MbConv_Block, self).__init__(**kwargs)
+
+        self.bn = keras.layers.BatchNormalization()
+        self.act = tfa.layers.GELU()
+
+        if conv_shortcut:
+            if strides > 1:
+                self.conv_shortcut_pool = keras.layers.AveragePooling2D(
+                    pool_size=(strides, strides), padding="same"
+                )
+            self.conv_shortcut_layer = keras.layers.Conv2D(
+                output_channel, kernel_size=1, strides=1, padding="same", use_bias=False
+            )
+
+        self.conv_1 = (
+            keras.layers.Conv2D(
+                filters=input_channel * expansion,
+                kernel_size=(1, 1),
+                strides=(1, 1),
+                padding="same",
+                use_bias=False,
+            ),
+        )
+        self.bn_1 = keras.layers.BatchNormalization()
+        self.act_1 = tfa.layers.GELU()
+        self.dconv_1 = (
+            keras.layers.DepthwiseConv2D(
+                kernel_size=(3, 3),
+                strides=(strides, strides),
+                padding="same",
+                use_bias=False,
+            ),
+        )
+        self.bn_2 = keras.layers.BatchNormalization()
+        self.act_2 = tfa.layers.GELU()
+
+        self.se_module = SE_Module(
+            input_channel * expansion, ratio=se_ratio / expansion, activation="gelu"
+        )
+
+        self.conv_2 = keras.layers.Conv2D(
+            filters=output_channel,
+            kernel_size=(3, 3),
+            strides=(1, 1),
+            padding="same",
+            use_bias=False,
+        )
+        self.drop_1 = keras.layers.Dropout(rate=dropout_rate)
+
+    def __make_divisible(self, value, divisor, min_value=None):
+        """Refer: https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet/mobilenet.py"""
+        if min_value is None:
+            min_value = divisor
+        new_value = max(min_value, int(value + divisor / 2) // divisor * divisor)
+        # Make sure that round down does not go down by more than 10%.
+        if new_value < 0.9 * value:
+            new_value += divisor
+        return new_value
+
+    def call(self, inputs):
+        x = self.bn(inputs)
+        x = self.act(x)
+        pre_act = x
+
+        if hasattr(self, "conv_shortcut_layer"):
+            if hasattr(self, "conv_shortcut_pool"):
+                pre_act = self.conv_shortcut_pool(pre_act)
+            shortcut = self.conv_shortcut_layer(pre_act)
+        else:
+            shortcut = inputs
+
+        x = self.conv_1[0](x)
+        x = self.bn_1(x)
+        x = self.act_1(x)
+        x = self.dconv_1[0](x)
+        x = self.bn_2(x)
+        x = self.act_2(x)
+        x = self.se_module(x)
+        x = self.conv_2(x)
+        x = self.drop_1(x)
+        x = keras.layers.Add()([x, shortcut])
+        return x
```

Comment:
Can't we make the class a function instead? That way the end model becomes fully expandable.

If we wrap a block like this inside a `layers.Layer` then we also lose access to its internals. So, prefer making this block with the Functional API. Here's an example:

https://github.com/keras-team/keras/blob/f95c9050ae577d97f2a7cae84a3096bbc9a1d869/keras/applications/convnext.py#L238-#L297

Reply:
Cool! Will try to implement it using Functional API.

Reply:
@sayakpaul If I am not wrong this is [MBConv](https://github.com/keras-team/keras-applications/blob/06fbeb0f16e1304f239b2296578d1c50b15a983a/keras_applications/efficientnet.py#L119) Block right?

Reply:
Yes, I think so. We can reuse it with due credits. @gustheman WDYT?

Reply:
If @gustheman is okay with it. I will modify the code and make the changes.